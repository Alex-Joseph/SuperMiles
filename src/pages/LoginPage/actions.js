import { actionTypes } from './constants';


export const authenticateUser = (username, password, history) => (dispatch) => {
  if (username === 'username' && password === 'password') {
    dispatch(loginSuccess())
    history.push('/');
  } else {
    dispatch(loginFailure())
  }
};

export const logoutUser = history => (dispatch) => {
  history.push('/login');
  dispatch(logout())
};

export const loginSuccess = () => ({
  type: actionTypes.LOGIN_SUCCESS,
});

export const loginFailure = () => ({
  type: actionTypes.LOGIN_FAILURE,
});

export const logout = () => ({
  type: actionTypes.LOGOUT,
});
