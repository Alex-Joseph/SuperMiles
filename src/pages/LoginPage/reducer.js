import { fromJS } from 'immutable';
import { actionTypes } from './constants';


export const initialState = fromJS({
  loggedIn: false,
  loginError: false,
});

export const authReducer = (state = initialState, action = {}) => {
  switch (action.type) {
    case actionTypes.LOGIN_SUCCESS:
      return state
        .set('loginError', false)
        .set('loggedIn', true);
    case actionTypes.LOGIN_FAILURE:
      return state
        .set('loginError', true);
    case actionTypes.LOGOUT:
      return state
        .set('loggedIn', false);
    default:
      return state;
  }
};
