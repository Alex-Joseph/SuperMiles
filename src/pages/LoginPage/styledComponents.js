import styled from 'styled-components';


const StyledLoginWrapper = styled.div`
  display: grid;
  place-content: center;
  padding: 24px 12px;
  grid-template-columns: 350px;
`;

export { StyledLoginWrapper };
