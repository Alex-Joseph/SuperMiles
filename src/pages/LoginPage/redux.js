import { authenticateUser } from './actions';


export const mapStateToProps = state => ({
  loggedIn: state.toJS().auth.loggedIn,
  loginError: state.toJS().auth.loginError,
});

export const mapDispatchToProps = {
  authenticateUser,
};
