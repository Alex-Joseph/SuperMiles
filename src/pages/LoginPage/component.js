import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { Button, Form, Message } from 'semantic-ui-react';
import { StyledLoginWrapper } from './styledComponents';
import Navbar from '../../components/Navbar';


class LoginPage extends Component {
  state = {
    username: '',
    password: '',
  }

  onChangeUserName = ev => this.setState({ username: ev.target.value });

  onChangePassword = ev => this.setState({ password: ev.target.value });

  handleSubmit = () => {
    const { username, password } = this.state;
    this.props.authenticateUser(username, password, this.props.history);
  }

  render() {
    const { loginError } = this.props;

    return (
      <Fragment>
        <Navbar title="SuperMiles" />
        <StyledLoginWrapper>
          <Form error={loginError} onSubmit={this.handleSubmit}>
            <Form.Input
              label='Username'
              value={this.state.username}
              onChange={this.onChangeUserName}
              type='text'
              maxLength={50}
              autoComplete='username'
            />
            <Form.Input
              label='Password'
              type='password'
              value={this.state.password}
              onChange={this.onChangePassword}
              maxLength={30}
              autoComplete='current-password'
            />
            <Message
              error
              header='Login Failed'
              content='please check your username and password'
              />
            <Button type="submit">Submit</Button>
          </Form>
        </StyledLoginWrapper>
      </Fragment>
    )
  }
}

LoginPage.propTypes = {
  loginError: PropTypes.bool.isRequired,
};

export default LoginPage
