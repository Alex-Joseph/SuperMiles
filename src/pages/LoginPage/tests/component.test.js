import React from 'react';
import { shallow, mount } from 'enzyme';
import LoginPage from '../component';


describe('LoginPage', () => {
  it('renders correct snapshot', () => {
    const wrapper = shallow(
      <LoginPage
        loggedIn={false}
        loginError={false}
        history={{}}
        authenticateUser={jest.fn()}
      />
    );
    const expectedLabel = 'Username';
    const expectedType = 'password';
    const formInputs = wrapper.find('FormInput');

    expect(wrapper.find('Form').props().error).toBe(false);
    expect(formInputs).toHaveLength(2);
    expect(formInputs.first().props().label).toEqual(expectedLabel);
    expect(formInputs.last().props().type).toEqual(expectedType);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders an error in the form if incorrect credentials are submitted', () => {
    const wrapper = shallow(
      <LoginPage
        loggedIn={false}
        loginError={true}
        history={{}}
        authenticateUser={jest.fn()}
      />
    );

    const expectedHeader = 'Login Failed';

    expect(wrapper.find('Form').props().error).toBe(true);
    expect(wrapper.find('Message').props().header).toEqual(expectedHeader);
  });

  it('calls authenticateUser with username and password onSubmit', () => {
    const authenticateUser = jest.fn();
    const wrapper = mount(
      <LoginPage
        history={{}}
        loggedIn={false}
        loginError={true}
        authenticateUser={authenticateUser}
      />
    );
    const username = 'user1';
    const password = 'password';

    wrapper.setState({ username, password });
    wrapper.find('[type="submit"]').at(0).simulate('submit');

    expect(authenticateUser).toHaveBeenCalledWith(username, password, {});
  });
});
