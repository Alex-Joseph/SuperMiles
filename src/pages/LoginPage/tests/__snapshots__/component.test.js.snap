// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage renders correct snapshot 1`] = `
<Fragment>
  <Navbar
    title="SuperMiles"
  />
  <styled.div>
    <Form
      as="form"
      error={false}
      onSubmit={[Function]}
    >
      <FormInput
        as={[Function]}
        autoComplete="username"
        control={[Function]}
        label="Username"
        maxLength={50}
        onChange={[Function]}
        type="text"
        value=""
      />
      <FormInput
        as={[Function]}
        autoComplete="current-password"
        control={[Function]}
        label="Password"
        maxLength={30}
        onChange={[Function]}
        type="password"
        value=""
      />
      <Message
        content="please check your username and password"
        error={true}
        header="Login Failed"
      />
      <Button
        as="button"
        type="submit"
      >
        Submit
      </Button>
    </Form>
  </styled.div>
</Fragment>
`;
